<a href="https://colab.research.google.com/github/anduc146khmt/DataAnalyst_IDBM_movies/blob/master/IDMB_movies.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

As we have a basic understanding of the different data structures in Pandas, let’s explore the fun and and get our hands dirty by performing practical data analysis on real data.

Here's a data set of 1,000 most popular movies on IMDB in the last 10 years . The data included  fields: Title, Genre, Description, Director, Actors, Year, Runtime, Rating, Votes, Revenue, Metascrore

It is an real dataset of kaggle community,  you can download it from [IMDB data from 2006 to 2016](https://www.kaggle.com/PromptCloudHQ/imdb-data)

Feel free to tinker with it and derive interesting insights.



#Usage

This project could be used to revise your knowledge or warming up your data analyst skill using Pandas

# Requirement


*   Having background about Python Programming, basic Data Analyst using Pandas
*   A little knowledge about Data Visualization using Seaborn



#Contributor

This notebook is my project practicing Pandas after finishing the Pandas course in kaggle community. I think it is a good course to begin learning Data Analyst, I recommend this course here [Pandas Course](https://www.kaggle.com/learn/pandas)

Author : [anduc146khmt](https://github.com/anduc146khmt)

#1. Read Data 

**Import Pandas and Seaborn**

In [5]:
import pandas as pd
import seaborn as sb

**Read CSV**

In [9]:
path = '/content/sample_data/IMDB-Movie-Data.csv'
data = pd.read_csv(path)

# 2. View Data

In [10]:
data.head()

Rank                    Title  ... Revenue (Millions) Metascore
0     1  Guardians of the Galaxy  ...             333.13      76.0
1     2               Prometheus  ...             126.46      65.0
2     3                    Split  ...             138.12      62.0
3     4                     Sing  ...             270.32      59.0
4     5            Suicide Squad  ...             325.02      40.0

[5 rows x 12 columns]

#3. Understand some basic information about the data

**Data Info**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


**Data Describe**

In [ ]:
data.describe()

Rank         Year  ...  Revenue (Millions)   Metascore
count  1000.000000  1000.000000  ...          872.000000  936.000000
mean    500.500000  2012.783000  ...           82.956376   58.985043
std     288.819436     3.205962  ...          103.253540   17.194757
min       1.000000  2006.000000  ...            0.000000   11.000000
25%     250.750000  2010.000000  ...           13.270000   47.000000
50%     500.500000  2014.000000  ...           47.985000   59.500000
75%     750.250000  2016.000000  ...          113.715000   72.000000
max    1000.000000  2016.000000  ...          936.630000  100.000000

[8 rows x 7 columns]

**Data Shape**

In [ ]:
data.shape

(1000, 12)

#4. Data Selection – Indexing and Slicing data

**Data Selection by columns**

In [ ]:
cols = ['Title', 'Genre', 'Actors', 'Director', 'Rating']
some_cols = data[cols]
some_cols.head()

Title  ... Rating
0  Guardians of the Galaxy  ...    8.1
1               Prometheus  ...    7.0
2                    Split  ...    7.3
3                     Sing  ...    7.2
4            Suicide Squad  ...    6.2

[5 rows x 5 columns]

**Data Selection using loc**

Now, I would select some information of the movie "Suicide Squad". And I have to do some following tasks

**Create new DataFrame with Titile column**

In [ ]:
path = '/content/sample_data/IMDB-Movie-Data.csv'
title_data = pd.read_csv(path, index_col = 'Title')
title_data.head()

**Select query movie from new DataFrame**

In [ ]:
index = ['Suicide Squad']
cols = ['Genre', 'Actors', 'Director', 'Rating', 'Revenue (Millions)']
suicide_squad = title_data.loc[index, cols]
suicide_squad.head()

Genre  ... Revenue (Millions)
Title                                    ...                   
Suicide Squad  Action,Adventure,Fantasy  ...             325.02

[1 rows x 5 columns]

**Select some optional record from dataset**

In [ ]:
indices = [0,5,10,15,20,25]
cols = ['Title', 'Genre', 'Actors', 'Director', 'Rating']
query_record = data.loc[indices, cols]
query_record.head()

Title  ... Rating
0                   Guardians of the Galaxy  ...    8.1
5                            The Great Wall  ...    6.1
10  Fantastic Beasts and Where to Find Them  ...    7.5
15                  The Secret Life of Pets  ...    6.6
20                                     Gold  ...    6.7

[5 rows x 5 columns]

**Data Selection using iloc**

In [ ]:
cols = ['Title', 'Genre', 'Actors', 'Director', 'Rating']
block_record = data.iloc[10:15][cols]
block_record.head()

#5. Data Selection – Based on Conditional filtering

Now, I would like to select some movie which has some conditions : 


1.   Movie are released from 2010 to 2016
2.   Genre of movie is Adventure or Sci-Fi
3.   Rating point larger than 6
4.   Movie are topped in terms of revenue   

In [ ]:
query_movie = data.loc[
      ((data['Year'] >=  2010) &  (data['Year'] <=  2016)) 
      &  data['Genre'].str.contains('Adventure|Sci-Fi', regex = True)
      &   (data['Rating'] > 6.0)  
      &  (data['Revenue (Millions)'] > data['Revenue (Millions)'].quantile(0.95))
]
cols = ['Title', 'Genre', 'Actors', 'Director', 'Rating']
query_movie = query_movie.loc[ : , cols]
print(' We have', len(query_movie.index), 'movies meet the requirements')
query_movie.head()

 We have 30 movies meet the requirements


Title                       Genre  ...        Director Rating
0   Guardians of the Galaxy     Action,Adventure,Sci-Fi  ...      James Gunn    8.1
4             Suicide Squad    Action,Adventure,Fantasy  ...      David Ayer    6.2
12                Rogue One     Action,Adventure,Sci-Fi  ...  Gareth Edwards    7.9
15  The Secret Life of Pets  Animation,Adventure,Comedy  ...    Chris Renaud    6.6
33                 Deadpool     Action,Adventure,Comedy  ...      Tim Miller    8.0

[5 rows x 5 columns]

#6. Groupby operations

Now, I want to know the average rating through all movies of directors. In this case, I would use the power of **groupby()**

In [ ]:
average_rating = data.groupby('Director').Rating.mean()
average_rating

Director
Aamir Khan             8.50
Abdellatif Kechiche    7.80
Adam Leon              6.50
Adam McKay             7.00
Adam Shankman          6.30
                       ... 
Xavier Dolan           7.55
Yimou Zhang            6.10
Yorgos Lanthimos       7.20
Zack Snyder            7.04
Zackary Adler          5.10
Name: Rating, Length: 644, dtype: float64

#7. Create DataFrame With Optional Requirements


Also, I would like to count the number of movies in each genre and display information as a DataFrame

In [ ]:
genres = ['Action', 'Adventure', 'Sci-Fi', 'Comedy', 'Fantasy', 'Animation', 'Thriller', 'Horror', 'Mystery' ]
numOfGenre = []
for index in range(len(genres)):
      n_genre = data.Genre.map(lambda genre:  genres[index] in  genre).sum()
      numOfGenre.append(n_genre)
genreDataFrame = pd.DataFrame(numOfGenre, index = genres, columns = ['Genres'])
genreDataFrame

Genres
Action        303
Adventure     259
Sci-Fi        120
Comedy        279
Fantasy       101
Animation      49
Thriller      195
Horror        119
Mystery       106

#8. Sorting operation

In this section, I would like to know which movie has the best rating and its descending order, alongside with its director.

In [37]:
cols = ['Title', 'Director']
sort_data = data.groupby(cols)[['Rating']].mean().sort_values(by = ['Rating'],  ascending = False)
sort_data.head(15)

Rating
Title                 Director                                
The Dark Knight       Christopher Nolan                    9.0
Dangal                Nitesh Tiwari                        8.8
Inception             Christopher Nolan                    8.8
Interstellar          Christopher Nolan                    8.6
The Intouchables      Olivier Nakache                      8.6
Kimi no na wa         Makoto Shinkai                       8.6
Taare Zameen Par      Aamir Khan                           8.5
The Lives of Others   Florian Henckel von Donnersmarck     8.5
The Prestige          Christopher Nolan                    8.5
The Departed          Martin Scorsese                      8.5
The Dark Knight Rises Christopher Nolan                    8.5
Whiplash              Damien Chazelle                      8.5
Django Unchained      Quentin Tarantino                    8.4
WALL·E                Andrew Stanton                       8.4
3 Idiots              Rajkumar Hirani                      8.4

#9. Dealing with missing values

Let's return **data.info()**, you have seen  Revenue (Millions) with  872 non-null, and Metascore  with 936 non-null and others is 1000 non-null. Why ?

It just because in these fields the data is missing, some data cells did not contain any values. So we will cleaning data by using some following steps

**Checking data is missing**

In [ ]:
revenue_null = data[pd.isnull(data['Revenue (Millions)'])]
metascope_null = data[pd.isnull(data['Metascore'])]
data_null = pd.concat([revenue_null, metascope_null])
data_null

Rank                    Title  ... Revenue (Millions) Metascore
7       8                 Mindhorn  ...                NaN      71.0
22     23           Hounds of Love  ...                NaN      72.0
25     26          Paris pieds nus  ...                NaN       NaN
39     40                5- 25- 77  ...                NaN       NaN
42     43  Don't Fuck in the Woods  ...                NaN       NaN
..    ...                      ...  ...                ...       ...
967   968                 The Walk  ...              10.14       NaN
969   970          The Lone Ranger  ...              89.29       NaN
971   972                Disturbia  ...              80.05       NaN
989   990                    Selma  ...              52.07       NaN
992   993     Take Me Home Tonight  ...               6.92       NaN

[192 rows x 12 columns]

**Replace data missing with particular values. Personally, I will replace missing data with 0**

In [ ]:
data_null.fillna(0)

Rank                    Title  ... Revenue (Millions) Metascore
7       8                 Mindhorn  ...               0.00      71.0
22     23           Hounds of Love  ...               0.00      72.0
25     26          Paris pieds nus  ...               0.00       0.0
39     40                5- 25- 77  ...               0.00       0.0
42     43  Don't Fuck in the Woods  ...               0.00       0.0
..    ...                      ...  ...                ...       ...
967   968                 The Walk  ...              10.14       0.0
969   970          The Lone Ranger  ...              89.29       0.0
971   972                Disturbia  ...              80.05       0.0
989   990                    Selma  ...              52.07       0.0
992   993     Take Me Home Tonight  ...               6.92       0.0

[192 rows x 12 columns]

#10. Dropping columns and null values

Another way to deal with null values is to delete cells with contains null values

In [22]:
data.dropna()

Rank                     Title  ... Revenue (Millions) Metascore
0       1   Guardians of the Galaxy  ...             333.13      76.0
1       2                Prometheus  ...             126.46      65.0
2       3                     Split  ...             138.12      62.0
3       4                      Sing  ...             270.32      59.0
4       5             Suicide Squad  ...             325.02      40.0
..    ...                       ...  ...                ...       ...
993   994  Resident Evil: Afterlife  ...              60.13      37.0
994   995                 Project X  ...              54.72      48.0
996   997           Hostel: Part II  ...              17.54      46.0
997   998    Step Up 2: The Streets  ...              58.01      50.0
999  1000                Nine Lives  ...              19.64      11.0

[838 rows x 12 columns]

As you see, it only contains **838 rows**, **dropna()** is used to drop rows which has null value in its cell.

Optional, you can choose to drop rows by set **axis = 0**, and columns with **axis = 1** (default parameter is **axis = 0**)

Another example, I see that I don't use **Runtime (Minutes)** and **Votes** for my analyst, I care more about rating, genre, revenue,... so that I will drop two columns **Runtime (Minutes)**  and **Votes**

In [26]:
drop_elements = ['Runtime (Minutes)',  'Votes']
data.drop(drop_elements , axis = 1)

Rank                    Title  ... Revenue (Millions) Metascore
0       1  Guardians of the Galaxy  ...             333.13      76.0
1       2               Prometheus  ...             126.46      65.0
2       3                    Split  ...             138.12      62.0
3       4                     Sing  ...             270.32      59.0
4       5            Suicide Squad  ...             325.02      40.0
..    ...                      ...  ...                ...       ...
995   996     Secret in Their Eyes  ...                NaN      45.0
996   997          Hostel: Part II  ...              17.54      46.0
997   998   Step Up 2: The Streets  ...              58.01      50.0
998   999             Search Party  ...                NaN      22.0
999  1000               Nine Lives  ...              19.64      11.0

[1000 rows x 10 columns]

# 11. Apply( ) functions

After analysting data, I would like to evaluate each movies by following this rules:  

*   If rating is larger than 8.5, the movie is reviewed "Excellent" 
*   If rating in range [7, 8.5), the movie is reviewed "Good"
*   If rating in range [5, 7), the movie is reviewed "Average"
*  If rating in smaller than 5, the movie is reviewed "Bad"



In [35]:
def evaluate (rating):
      if rating >= 8.5:
        return "Excellent"
      elif rating >= 7 and rating < 8.5:
        return "Good"
      elif rating >= 5 and rating < 7:
        return "Average"
      else:
          return "Bad"
data['Evaluation'] = data['Rating'].apply(evaluate)
index = ['Title', 'Genre', 'Director', 'Rating', 'Evaluation', 'Description']
data_query = data.loc[ : , index]
data_query.head()

Title  ...                                        Description
0  Guardians of the Galaxy  ...  A group of intergalactic criminals are forced ...
1               Prometheus  ...  Following clues to the origin of mankind, a te...
2                    Split  ...  Three girls are kidnapped by a man with a diag...
3                     Sing  ...  In a city of humanoid animals, a hustling thea...
4            Suicide Squad  ...  A secret government agency recruits some of th...

[5 rows x 6 columns]

You also can use sorted data **(in 8th section)** and categorize movies from the highest-rating movie to lowest-rating movie.

In [38]:
sort_data['Evaluation'] = sort_data['Rating'].apply(evaluate)
sort_data

Rating Evaluation
Title                   Director                            
The Dark Knight         Christopher Nolan     9.0  Excellent
Dangal                  Nitesh Tiwari         8.8  Excellent
Inception               Christopher Nolan     8.8  Excellent
Interstellar            Christopher Nolan     8.6  Excellent
The Intouchables        Olivier Nakache       8.6  Excellent
...                                           ...        ...
Wrecker                 Micheal Bafaro        3.5        Bad
Tall Men                Jonathan Holbrook     3.2        Bad
Dragonball Evolution    James Wong            2.7        Bad
Don't Fuck in the Woods Shawn Burkett         2.7        Bad
Disaster Movie          Jason Friedberg       1.9        Bad

[1000 rows x 2 columns]

# Conclusion

Well, thank you for your reading, I hope you have learned something through my projects.

If you see this notebook is useful, I very glad if you give me a star in my github projects.

See more my projects here : [https://github.com/anduc146khmt](https://github.com/anduc146khmt)

Contact me : ducan1406@gmail.com